core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -0,0 +1,533 @@
+        getOrCreatePool(actualProtocol, key).addLast(pooledChannel);
+    }
+
+    /**
+     * Closes all {@link Channel}s managed by this pool.
+     */
+    @Override
+    public void close() {
+        closed = true;
+
+        if (eventLoop.inEventLoop()) {
+            // While we'd prefer to block until the pool is actually closed, we cannot block for the channels to
+            // close if it was called from the event loop or we would deadlock. In practice, it's rare to call
+            // close from an event loop thread, and not a main thread.
+            doCloseAsync();
+        } else {
+            doCloseSync();
+        }
+    }
+
+    private void doCloseAsync() {
+        if (allChannels.isEmpty()) {
+            return;
+        }
+
+        final List<ChannelFuture> closeFutures = new ArrayList<>(allChannels.size());
+        for (Channel ch : allChannels.keySet()) {
+            // NB: Do not call close() here, because it will trigger the closeFuture listener
+            //     which mutates allChannels.
+            closeFutures.add(ch.closeFuture());
+        }
+
+        closeFutures.forEach(f -> f.channel().close());
+    }
+
+    private void doCloseSync() {
+        final CountDownLatch outerLatch = eventLoop.submit(() -> {
+            if (allChannels.isEmpty()) {
+                return null;
+            }
+
+            final int numChannels = allChannels.size();
+            final CountDownLatch latch = new CountDownLatch(numChannels);
+            if (numChannels == 0) {
+                return latch;
+            }
+            final List<ChannelFuture> closeFutures = new ArrayList<>(numChannels);
+            for (Channel ch : allChannels.keySet()) {
+                // NB: Do not call close() here, because it will trigger the closeFuture listener
+                //     which mutates allChannels.
+                final ChannelFuture f = ch.closeFuture();
+                closeFutures.add(f);
+                f.addListener((ChannelFutureListener) future -> latch.countDown());
+            }
+            closeFutures.forEach(f -> f.channel().close());
+            return latch;
+        }).syncUninterruptibly().getNow();
+
+        if (outerLatch != null) {
+            boolean interrupted = false;
+            while (outerLatch.getCount() != 0) {
+                try {
+                    outerLatch.await();
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                }
+            }
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    static final class PoolKey {
+        final String host;
+        final String ipAddr;
+        final int port;
+        final int hashCode;
+
+        PoolKey(String host, String ipAddr, int port) {
+            this.host = host;
+            this.ipAddr = ipAddr;
+            this.port = port;
+            hashCode = ipAddr.hashCode() * 31 + port;
```

Comment:
Is it ok to use `host` in `equals` but not in `hashCode`? If it is let's add a comment since it's not conventional.

Reply:
It's OK, but I guess it doesn't do harm by adding `host.hashCode()`, so I added it. :-)

Reply:
That is, without `host.hashCode()` it had higher chance of collision for the servers that utilize virtual hosts.